### ISSUE TITLE
Publish v5 draft

### ISSUE THREAD

**Issue by awwright** [json-schema/json-schema#167]
The v4 draft is expired, and people are still using it, and this is bad for some arbitrary reason that is bad enough the IETF still has an expiration system for Internet-drafts. I'd like to publish draft-05 within two month's time.

I propose doing the following things:

(1) The json-schema.org website should really get its own repository all to itself. Call it "json-schema.org" maybe

(2) I would expect current work to be on the `master` branch, so put current work there

(3) Tag commits that drafts are/were published against

(4) Create a "draft-05" milestone and label the relevant issues that we want to solve. Insert editor's notes for the other issues that will likely see changes in v6, but that we can't address within two months time.

If in doubt, copy https://github.com/httpwg since they seem to know what they're doing

(5) I'd like to see the master branch change structure to the following layout:

* archive/draft-$number/{core,schema,hyper-schema}.xml - the XML sources for previous, published drafts that we need reverse-comparability with
* archive/draft-$number/{schema,hyper-schema}.json - the meta-schemas themselves
* {core,schema,hyper-schema}.xml - current work on next draft to-be-published
* Makefile - calls xml2rfc to build `%.html` and `%.txt` from `%.xml`
* README - read this first

I've made a new repo from scratch along these lines, but maybe we can just rearrange the existing one, except I have no clue which file is 'current' and which files are 'archive'.

**Comment by Julian** [comment]
+1

**Comment by Fannon** [comment]
+1

**Comment by tmarrs** [comment]
+1

**Comment by epoberezkin** [comment]
Is there a list of changes for v5?

**Comment by Acubed** [comment]
Part of this task is sorting through the current list of issues and tagging milestones.

I don't think there's a formal list of issues, but fixing incorrect vocabulary usage would top my list.

**Comment by epoberezkin** [comment]
So there are no actual changes for validation keywords?

**Comment by Acubed** [comment]
@epoberezkin No one's started any work yet, this is just my feature request to get an active (unexpired) JSON Schema draft published

**Comment by epoberezkin** [comment]
Got it! Thanks

**Comment by donaldpipowitch** [comment]
I would love to see [switch](https://github.com/json-schema/json-schema/wiki/switch-(v5-proposal)) and [constant](https://github.com/json-schema/json-schema/wiki/constant-(v5-proposal)) in v5.

**Comment by Acubed** [comment]
@donaldpipowitch The goal for this issue is to clean up outright errors and internal contradictions in the text, and to release a once-again-unexpired draft.

If those aren't already issues in the tracker, elsewhere, you should create a new one.

**Comment by donaldpipowitch** [comment]
With "issues in the tracker" do you mean issues in this Git repo?

**Comment by Relequestual** [comment]
Is there any clear leadership on this project? There's no infomation I can see that explains how to get involved.

**Comment by kriszyp** [comment]
I started the JSON Schema specification, and Gary Court helped shape and 
author the first few versions. I stepped back, and Francis Galiegue led the 
work on v4 (with help from others), and has since stepped back. Nick Lombard 
owns the json-schema github organization. If someone wants to step up to 
lead future efforts, they certainly are welcome.
Thanks,
Kris

**Comment by Relequestual** [comment]
@nickl- Hey! So Kris, Gary, and Francis stepped back from working on json-schema. Are there any leads apart from yourself owning the org here on github?

**Comment by Clemens-U** [comment]
Any progress on keep JSON schema specification going? As I already mentioned on other places I can manage the json-schema.org web pages if there is demand (@kriszyp).

**Comment by Relequestual** [comment]
There's a number of hyper-schema libaries / packages / modules for various languages now. Only a few are listed. It's clear updates would be benifical! =] Also happy to help out!

**Comment by geraintluff** [comment]
Kris and Gary passed it on to Francis, who passed it on to me.  I'm trying to carve out more time for it, but it's been difficult to find the time.

(Although Nick Lombard created the organisation, he's not been involved with any of the specs - he handled some of the GitHub stuff.)

**Comment by epoberezkin** [comment]
I could help too. At the very least I would like to get involved in the discussion about standard changes.

**Comment by jasoniangreen** [comment]
I can help too. I work with @epoberezkin and given that we are making extensive use of json-schema in our current role, we can definitely find time for it.

**Comment by epoberezkin** [comment]
@geraintluff why don't you add couple people as collaborators, we could review pull requests at least?

**Comment by geraintluff** [comment]
Unfortunately, one of the problems is that since we (Francis and I and Kris and Gary) got locked out of the GitHub organisation (shortly before Nick Lombard disappeared over a year ago), we don't actually have the permissions on this organisation to add collaborators.

In view of the proposed changes (which sound generally positive to me), and the above issue, I think we should consider moving to a new organisation.

**Comment by Clemens-U** [comment]
"... I think we should consider moving to a new organisation" Sounds like the only way to go if nobody can reach Nick anymore.

**Comment by geraintluff** [comment]
OK, so if we wanted to do that, I think some steps might be like:

* create new organisation, following something like @Acubed's proposed repo structure.  I'm happy to do this, but am open to help (particularly if @Acubed already has something)
* collect people interested in various aspects (site, tests, spec discussion), permission them up appropriately
* each group agrees on change/review policy, how much consensus required to accept pull requests etc.
* once things are stable, @kriszyp points the json-schema.org domain at the new repo, and we replace the existing repos with notifications of the move.
* discuss v5 and publish updated spec(s)

For me: I'm very invested in the spec, and would love to continue collecting proposals and feedback on them etc.  I'm also happy to continue editing the spec (I did the hyper-schema part for v4, and can extend both), but would prefer the spec discussions to be clearly separate from nailing down what's already been agreed.

I think with more people on board this could start moving quickly again - and actually being able to grant permissions for repos so helpful people can be effective would obviously be a big step in that. :relieved:

**Comment by Julian** [comment]
Eh, if we haven't already, we should reach out to GitHub @support -- I
don't know if they'll help or say they can't, but it's worth a shot.

**Comment by Julian** [comment]
(Which I've just done. Will update back if they help.)

**Comment by geraintluff** [comment]
Cool - thanks. :)

I think that even if we get control of the organisation back, a "clean slate" approach has some benefits (splitting up repos etc.).

**Comment by Julian** [comment]
It has some I agree, although we could certainly split up repos anyhow, and
try to not break bookmarks or search results as we do it, which would kind
of confuse.

**Comment by Relequestual** [comment]
Agreed with all of the above. Hopefully github support are gracious and have space in their policy and procedures for events like this. Getting back control of the github org would be benificial in making clear the cononical repositories.

/sidebar
I'm invested in json-schema, and I'd like to become more so. I'm involved in a group called The Global Alliance for Genomic Health, and we recetly identified AVRO not really fit for our purpose. I'd like to be able to push json-schema and json hyper-schema forward as a solution.

Hopefully this can lead to a number of key and highly skilled individuals from the GA4GH coming to help out with this project! =]
/

I haven't been involved in this sort of work before, but I'm keen to learn. Probably best suited to helping out with the site and documentation, but woud like to learn how I can contirbute to the specification also.

**Comment by geraintluff** [comment]
@Julian - from what I've seen, when you rename a repo then GitHub sets up forwarding, so visitors get a 303 to the new location.

**Comment by Relequestual** [comment]
@geraintluff I'm seeing lots of questions regarding the status of json-schema etc. Do you have access to update the google group description to add a link to here maybe?

**Comment by geraintluff** [comment]
Sure - still an admin on the Google Group, that wasn't affected by the other issues.  I've added a link to the description.

**Comment by Julian** [comment]
Looks like the answer is "no" on whether we can get changes to the ownership settings.

**Comment by Acubed** [comment]
And @nickl- can't set ownership, for some reason? Hmm. You'd think they have some way to recover that.

---

> create new organisation, following something like @Acubed's proposed repo structure. I'm happy to do this, but am open to help (particularly if @Acubed already has something)

I think, ideally, we'd try to re-use the current org and repo. Having multiple "official" orgs could get confusing; and we'd lose the issues in the tracker (though I don't see this as a big an issue).

> collect people interested in various aspects (site, tests, spec discussion), permission them up appropriately
> each group agrees on change/review policy, how much consensus required to accept pull requests etc.

I was thinking for v5, we use the standard IETF consensus model (or unanimous consent). Which should be easy if we focus only on repo layout, and outright errors and internal contradictions in the text, as I propose.

> once things are stable, @kriszyp points the json-schema.org domain at the new repo, and we replace the existing repos with notifications of the move.

I think the GitHub repo itself declares the domain name it's using, not the other way around. I don't think there's anything that prevents multiple repos from declaring a single domain name. Though if I recall, you should still be able to remove the CNAME file from the repo.

> discuss v5 and publish updated spec(s)

(THUMB UP SIGN EMOJI)

I have a few edits to submit once the repo is organized and overall ready to accept changes.

I'm interested in helping with tagging issues, website, editing, etc. I'll be maintaining JSON Schema implementations indefinitely, and my interest in contributing is as an independent expert.

**Comment by donaldpipowitch** [comment]
Just a short question: What is the scope of this issue? "Just" renewing the JSON Schema spec to figure out the current ownership or also to add features discussed here: https://github.com/json-schema/json-schema/wiki/v5-Proposals?

**Comment by Relequestual** [comment]
@donaldpipowitch looks like it sort of evolved because of lack of activty.

@Julian Oh man, seriously? I find it difficult to believe they don't have procedures for such things. I'll try push a few buttons =/

**Comment by Relequestual** [comment]
@Acubed It's more that Nick isn't replying to any form of contact... =[ No one can get hold of him.

**Comment by Acubed** [comment]
@donaldpipowitch This issue proposes no feature changes, but as many bug fixes as we can get done in the timeframe. Things like updating references (RFC2616 is no longer current), fixing vocabulary (usage of URI vs URI-reference vs absolute-URI is mixed up), maybe making notes about possible future changes, etc.

**Comment by donaldpipowitch** [comment]
Thanks for clarifying this. I hope the RFC can soon be renewed and the ownership gets clear, so we can focus on new features (or adding the already discussed features).

**Comment by geraintluff** [comment]
@Acubed 

> I was thinking for v5, we use the standard IETF consensus model (or unanimous consent). Which should be easy if we focus only on repo layout, and outright errors and internal contradictions in the text, as I propose.

It depends on the group - the spec group might agree something like this, yeah.  However, if someone makes a PR to add a new implementation, then the web group presumably does not need unanimous consent to merge it. :)

> I think the GitHub repo itself declares the domain name it's using, not the other way around.

I can't remember quite how it works (I looked a while back but forgot), but I'd assumed there'd have to be agreement from both sides, otherwise forks would all end up competing for the domain.

**Comment by geraintluff** [comment]
@Julian - Thanks for talking to support.  So, that's a confirmed "no" on ownership?

In that case, I think we should go ahead with the new organisation.  It'll take a while to get things straight anyway, so we can keep trying if we like, and if we magically get ownership sorted in reasonable time, we can move everything back.

**Comment by Julian** [comment]
@geraintluff yup, confirmed.

I'm fine with proceeding with a new org I guess if it will keep the momentum here going, but I'll be hoping either @nickl- reappears or someone says a magic word sometime soon to save the trouble...

**Comment by geraintluff** [comment]
Yeah, I hope this gets resolved more sanely too.

For now at least, I've created https://github.com/json-schema-org.  We can start the refactor, and should get more people on board.

@Acubed, did you say you already had a repo?  If that's not around any more, would you be up for creating another one?

**Comment by Relequestual** [comment]
I've managed to get hold of Nicks previous employer, who provided his skype and phone number, which I have forwarded to @geraintluff. I can't try myself right now, but can't hurt to try at some point!

I'm putting myself up for taking on work on the website and spec. Suggest that there's a repo for logging issues / tasks to do with migration to the new org. Including post migration tasks like updating the website. I'm happy to help however I can. I have plenty web dev experience.

@geraintluff I wondered about the domain before now. It turns out, that the first repo to set the required file as the domain, gets made the domains repo. I guess github will deal with any contenst on a case by case basis. Do you have control over the domain btw?

**Comment by geraintluff** [comment]
@Relequestual - hero.  I'll try giving him a poke.

A migration/TTD repo sounds fairly reasonable.  Not sure if it should be a repo or a wiki, but AFAIU we can shift between the two.

The domain is in @kriszyp's hands.  We might not need to transfer anything if we're lucky, though.

**Comment by Relequestual** [comment]
I was mainly thinking a repo to allow the tasks to be audited and for work needed to be clearly laid out, allowing people to pick them up and work through them. Heck you guys all know this stuff =] I love audited processes!

**Comment by Acubed** [comment]
@geraintluff I have one, created from scratch, on my local system. To avoid confusion (and to avoid a "Forked from ..." message if it gets cloned to this org) I haven't published it yet, but it pretty much follows the layout I described.

I should probably also work on forking the website, too.

**Comment by Acubed** [comment]
@geraintluff I created https://github.com/Acubed/json-schema-spec

I'd think the best option is you create a new repo and push this, or I transfer ownership.

**Comment by richievos** [comment]
As a previous contributor who bowed out when the group seemed to be going a bit bonkers, I :+1: all the people in this thread getting involved.

**Comment by Relequestual** [comment]
OK, shall come up with ideas for a new github org name? https://github.com/json-schema/json-schema/issues/173

**Comment by darshakthakore** [comment]
Even I'm interested in contributing/helping with v5 of the JSON Schema. We are already using JSON Schemas in IoT modeling. Given that there's an IETF meeting in the next 2 weeks (July 18th to 24th), i wanted to check how many of the interested folks here will also be at the IETF?
Maybe we can meet up there ?

**Comment by brettz9** [comment]
Any chance localization will make it into a v5 spec?

**Comment by Acubed** [comment]
@brettz9 That'd be a topic for another issue, if you have a specific proposal.

This issue doesn't ask for new features, however: Just getting the existing issues and inconsistencies out of the way. This is in the interest of avoiding a sort of [X-Y problem](http://mywiki.wooledge.org/XyProblem), features being proposed and arguments occurring for no other reason than more fundamental errors/bugs/omissions that haven't been fixed yet.

**Comment by brettz9** [comment]
I was just referring to a proposal started earlier by @geraintluff at https://github.com/json-schema/json-schema/wiki/multilingual-meta-data-%28v5-proposal%29 . (I do have another approach in mind, but not sure whether to add it as a new wiki page or describe under an "Alternative approach" kind of heading.)

With the growing popularity of tools like [JSONEditor](https://github.com/jdorn/json-editor/), and i18n being important for projects of any significant size, I see this as quite a fundamental need and hope it may make it in, if not in the next version, then for any subsequent version taking in new features.

So, should this be opened as a new issue and reference the wiki page?

**Comment by DimDragon** [comment]
Hello, seems I'm bit late, but also willing to participate and help with what I can. Few words for myself:

I'm IT company CTO & co-owner ([ConsultCommerce LTD](http://consultcommerce.com)) and Technical/Project lead on big HR/Time & Attendance software ([ManusPlus](http://manusplus.com))

Last 2 years I'm involved with [HROpen Standards Consortium](http://hropenstandards.org) as part of technical team rebuilding specifications they provide as JSON Schema. As part of this project I have many use cases for current v4, along with some tough to cover.

I have also build JSON Schema v4 parser/validator/visualizer ([JSchema](http://jschema.com)). Project has evolved and we will be upgrading it in next 1-2 months

Finally I'm also working on JSON Hyper Schema extension specification to serve as documentation generation base for JSON REST APIs

I'm firm believer in JSON Schema and use it on daily basis, so will be honored to participate in improving it

**Comment by Relequestual** [comment]
Hey @DimDragon! That's a pretty impressive introduction there. Looks like you're involved in a lot of smart stuff!

Part of my proposal for moving the project forward includes forming a list of people who want to be involved, and organising everyone into teams. It sounds like you've got quite a bit of experience using json-schema in the real world, and that's really valuable.

I'll make sure you're involved in moving forward! =]

**Comment by tmarrs** [comment]
I should've responded sooner, but I'm heavily invested in JSON Schema these days, and I'd like to help.

Here are a few of my qualifications:
* I'm currently writing [JSON at Work](https://github.com/tmarrs/json-at-work) for O'Reilly. I cover Schema in chapter 5.
* I just delivered a 3 hour tutorail on [JSON Schema, Search, and Transform](https://github.com/tmarrs/presentations/tree/master/OSCON/2015/JSON-at-Work-Schema-Search-and-Transform) at OSCON 2015.
* I'm passionate about JSON Schema.

Please let me know where I can get plugged in and start helping. 
Thanks to everyone on this project for all you're doing to help and support the larger community.

**Comment by Acubed** [comment]
The top thing we need is reviews and +1s on all the issues/PRs open for https://github.com/json-schema-org/json-schema-spec/search?q=is%3Aopen&type=Issues&utf8=%E2%9C%93

**Comment by tmarrs** [comment]
Cool. I'll take a look.

**Comment by jmorey** [comment]
I am also interested in participating in the evolution of JSON Schema.  I'm involved in the specification of JSON based REST APIs at my current employer and in several standards groups. JSON Schema is heavily used/relied on in the specifications. I have an interest in JSON Schema being extended to cover common use cases that it does not currently work well for.

**Comment by yoshuawuyts** [comment]
Maybe it's a silly question: but what does it take to get JSON Schema to become an RFC? RFCs seem to be a big deal. Would be nice to consolidate things since pretty much everyone is using JSON Schema.

**Comment by Relequestual** [comment]
@yoshuawuyts Things are happening about this. I won't say exactly what, as I don't know if it is inteneded to be made public yet. I have emailed the relevent people and asked if they would pop in and update this issue when they can.

**Comment by Acubed** [comment]
@Relequestual Note that there's _very_ little that's not public about IETF proceedings. Virtually all of it is public.

Side note: Still waiting for comments on open issues at  https://github.com/json-schema-org/json-schema-spec !

**Comment by Relequestual** [comment]
Oh sure. I ment there are things happening currently to move the spec forward to being ready. But exactly what that is, isn't public knowlegde yet, and I don't feel I have the right to disclose that info. It's good things though! =D

**Comment by willcrick** [comment]
hi, i am a little confused about the status of this issue and this org in github, can someone clarify?
* The traditional website (which is what i always find from google) http://json-schema.org/ points to this repo still
* The linked issue about a new repo (which is closed so i assume its done) [mentions](https://github.com/json-schema/json-schema/issues/173#issuecomment-115690539) a new repo http://json-schema-org.github.io/ as set up but when i hit that URL it displays the same web page as above (but the link in it points back to this repo). Is the new repo just for the website, the spec or both?
* The new repo https://github.com/json-schema-org/json-schema-spec does not seem to have as much recent activity as this one. Is this one closed or still active? Do they both have a role or are we in the middle of a migration?

I can appreciate that its complicated to sort all this out, especially as an open source project, i just wanted to point out that its pretty confusing right now. It would be great to get v5 published if we can (happy to help in anyway if i can!).

**Comment by Acubed** [comment]
@willcrick We're having difficulty tracking down the actual owner of the org for this repository, so new work is being done at https://github.com/json-schema-org/json-schema-spec 

The top thing we need is reviews and +1s on all the issues/PRs open for https://github.com/json-schema-org/json-schema-spec/search?q=is%3Aopen&type=Issues&utf8=%E2%9C%93

**Comment by itdecapit** [comment]
Thank you for your work on JSON Schema! I hope it gets finalised and approved as some kind of official standard ASAP.

I have been lately involved in defining an integration architecture for a Corporation. Sadly I noticed being forced to demand that only SOAP/WS must be used. 

The issue is that it is not enough to build an ESB or similar service for connecting systems. When working with a multi-level security networks we have to produce some level of guarantee that *only* the approved messages are being delivered. For that we require schema validation (among other things) for all integrations done with the integration service.

Before JSON Schema has some kind of official standardization the commercial integration products will not implement it. Before that we can not even tailor a module, because it is too risky without an official standard. So, absurdly I have been forced to forbid all use of REST/JSON. This is very frustrating, because those are the technologies we should require.

Anyways, if there are janitorial tasks, or if you need assistance in the standardization process, you should probably set up an information wiki page for that. Not being native English speaker is a problem at least for my participation, but if there is something I could still do...

**Comment by Relequestual** [comment]
There's some work going on "behind the scenes" to bring draft 4 up to scratch so it can be made an official standard. Wonder if @geraintluff can give an update on that?

The whole "which repo is active" and "what's the deal with the website" are totally valid questions. I was hoping I'd be able to spend time on this project to fix some of those clear issues, but I simply haven't had the time to allocate that I thought I would, so apologies for that.

My intension was to create a new site from scratch (design wise) within the new repo. Realising I can't commit to that right now (or at least lack the motivation to get started), I think I need to re-focus on updating the current site.

My plan is to handling any outstanding pull requests on the repo found HERE. This includes updating the readme for the site and contributing.md to point to the new org and repo, and then work from there (after merging in any merged PRs). I think that sounds like a reasonable way to move forward.

**Comment by Relequestual** [comment]
@willcrick "difficulty tracking down the owner of this github org": We have tried, and failed, and failed, and github will allow anyone else to reclaim it. Eventually I guess we will be confident that everything is migrated, and then wipe the repos, having a md with a redirect link. Sucks though =[

**Comment by brettz9** [comment]
I had been hesitant to add a new proposal, but I figured that if you were voting on existing proposals, it would help to have relevant alternatives, so I have just now added my suggestion to the "Concerns" section of the existing multi-lingual proposal at https://github.com/json-schema/json-schema/wiki/multilingual-meta-data-(v5-proposal). I really think this alternative offers strong benefits--and i18n is really something I feel is sorely lacking in json-schema...

**Comment by willcrick** [comment]
ok thanks @Relequestual and @Acubed , i will take a look at the issues in the new repo then.

**Comment by willcrick** [comment]
p.s. if you want help with github process/organization in the new reop i can help

**Comment by JSaunders0102** [comment]
Is there an estimate when an unexpired version might be available (even if it's still v4)?
Thanks.

**Comment by timcharper** [comment]
I'm seeing that @kriszyp is still active on github. He's not responding, to anything? It would be ideal to get ownership of this project figured out because the google roads point here.

**Comment by kriszyp** [comment]
@timcharper I don't think I have any higher level access than @geraintluff does in this github project. It looks like we are both just members. Or do you see something that I don't?

**Comment by timcharper** [comment]
Yeah, sorry about that, please see my update :)

I emailed a friend of mine who works for Github and asked if he could get you assigned as owner for this organization. Hopefully, that should at least remove the log jam, for now.

**Comment by Relequestual** [comment]
@timcharper A few of us have been in contact with Github about this specific org. The only owner for this org is @nickl-, and he is not responding. Github will NOT change ownership for us dispite this. If you scroll up, and read the histroy of this issue, you will see this. (https://github.com/json-schema/json-schema/issues/167#issuecomment-114310963)

**Comment by epoberezkin** [comment]
@acubed @geraintluff @kriszyp @Relequestual etc. as it seems that the current json-schema organisation cannot change owners and the existing members would like to share the burden with others there seems to be no other option to move it to the other organisation you've created.

To begin with you could:
1) make several people who you trust the owners (so that the current situation is not repeated)
2) add some other people interested to contribute as members
3) fork all the repos from the current organisation to the new one (happy to do it when I am a member :).
4) point the existing domain to the new repo if possible, if not - create the new domain
5) put the notice(s) on this org (and the current domain if it can't be connected to the new repo) about the transition

I understand that the hope that somehow you can get the current organisation under control slows down these actions, but it seems like it is not going to happen any time soon. And even if it does you will have a choice to either switch back (if @nickl- becomes responsive soon) - we'll just PR the changes from new  org to the current one - or just keep it in deprecated state. But the standard will be moving forward instead of stagnating.

Once it is done it would be good to work out the mechanism of discussing the features that will be included in the version 5 of the standard, given that JSON-schema standard became quite valuable to the community.

I think it is very important to move the standard forward. Actually, I have already implemented all proposed v5 validation keywords and $data in [ajv](https://github.com/epoberezkin/ajv) validator; one of my PRs here is v5 meta schema. I also think it is equally important to preserve the JSON-schema design philosophy:

- "shallowness" - keywords can only depend on siblings and direct children
- "orthogonality" - low dependence of keywords on each other
- no/little overlapping of keywords
- leaving implementation details to implementations...

I am sure there are different opinions and I am happy to agree with any your decision, but I hope there will be a discussion of these new features and how they follow the existing (although unwritten) design principles, so that when these principles are changed it is done consciously with the full understanding of costs and implications.

I am really looking forward to some progress here and happy to help.

**Comment by sam-at-github** [comment]
> I also think it is equally important to preserve the JSON-schema design philosophy:
>
>    "orthogonality" - no/little overlapping of keywords
>    "independence" - low dependence of keywords on each other
>    "shallowness" - keywords can only depend on siblings and direct children ($data extends rather than violates it)
>    leaving implementation details to implementations...

Yep. Great insight. I agree its important to express this. 

However, I don't really get the hang up about not having owner privs on this repo and progression of this spec. It's overrated. At least, its premature. Implementation and standardization should go hand in hand, and I see a lack of implementation with respect to v5 features(?) - I do see divergence from the current spec where it's unimplementable or ambiguous though..

I'm also skeptical that github is a great format for driving development of specifications. The community feed back aspects are good, but that's already available here.

Can you pull together the changes you have made to AJV under "v5" in a formal and precise manner, preferably inline with the style of the existing spec, and stick on the wiki under "AJV v5 specification" or something? For what its worth I'd definitely take a look and implement it if it's well written. I think this sort of organic growth is the way forward for now.

**Comment by epoberezkin** [comment]
Ajv implements the following keywords:

- [switch](https://github.com/epoberezkin/ajv/blob/master/KEYWORDS.md#switch-v5-proposal)
- [contains](https://github.com/epoberezkin/ajv/blob/master/KEYWORDS.md#contains-v5-proposal) - only for a single item, I think that allowing multiple items in `contains` is a complication
- [constant](https://github.com/epoberezkin/ajv/blob/master/KEYWORDS.md#constant-v5-proposal)
- [patternGroups](https://github.com/epoberezkin/ajv/blob/master/KEYWORDS.md#patterngroups-v5-proposal) - I think it should complement rather than replace patternProperties and I also think that a simple `patternRequired` could be useful to make it similar to `properties`/`required`. I didn't implement it yet though.
- [formatMaximum, formatMinimum, exclusiveFormatMaximum, exclusiveFormatMinimum](https://github.com/epoberezkin/ajv/blob/master/KEYWORDS.md#formatmaximum--formatminimum-and-exclusiveformatmaximum--exclusiveformatminimum-v5-proposal).

It also implements [$data reference](https://github.com/json-schema/json-schema/wiki/$data-(v5-proposal)) with relative JSON-pointer as the possible value for the following keywords: constant, enum, format, maximum/minimum, exclusiveMaximum / exclusiveMinimum, maxLength / minLength, maxItems / minItems, maxProperties / minProperties, formatMaximum / formatMinimum, exclusiveFormatMaximum / exclusiveFormatMinimum, multipleOf, pattern, required, uniqueItems.

It provides [v5 meta-schema](https://github.com/epoberezkin/ajv/blob/master/lib/refs/json-schema-v5.json).

@sam-at-github did you mean wiki in the format / style of the [draft 4 document](http://json-schema.org/latest/json-schema-validation.html)?

The problem with the existing github org is that the current members don't have enough time and would like other people to participate (see above). And they can't add more members.

@geraintluff maybe an acceptable "temporary" solution would be to decide on additional "de-facto" members so that you or some other member can act as "proxies" merging PRs etc. based on confirmation from "de-facto" members in the issue without spending time to review them. What do you think about it?

**Comment by YurySk** [comment]
Hi All

Forgive me a dumb question. Is this a forum where I can comment on the draft v5 features? I have a decent amount of experience of generic xml processing on the enterprise scale, experience with building dev. tools and stuff like this. How moving heavily into Json.
Also, another dumb question. How does a governing body operates in json-schema. I.e. how decides what gets included and how.

Thank you,
Yury

**Comment by Relequestual** [comment]
@YurySk Probably best hit the Google Group, which you can find via the website, I think.

You're assuming the IS some sort of governing body... the reality is, there is not. As mentioned... somewhere in a thread on the GG or github, the hope is that an IETF working group will be able to pick it up and support it more than the people who currently can't give it as much love as everyone (including them) would like to.

**Comment by YurySk** [comment]
Thanks. I know about that group, just not seeing an ongoing feature discussion.

**Comment by sam-at-github** [comment]
> Ajv implements the following keywords:

@epoberezkin I'll have a look over these.  Any amendments? 

> @sam-at-github did you mean wiki in the format / style of the draft 4 document?

> The problem with the existing github org is that the current members don't have enough time and would like other people to participate (see above). And they can't add more members.

I mean with the same sentence structure etc, so its easy to later incorporate into the spec *later* and pickup for people who are familiar with the v4 spec. And so its fairly rigorous. Maybe you could fork a copy of this repo, make the changes you want in private, and c&p the relevant changes into a page on the wiki for AJV v5. 

I mean, suppose you were a collaborator on this project, what would you actually merge? You'd probably create a branch for draft v5, and update the spec with your set of changes. But there is no consensus on those changes yet. The way to build consensus and get weight behind your changes is by getting people to actually implement them. But creating a draft branch here, and decreeing it "v5" is premature, and won't work is what I'm saying. History has shown us it does not work.

**Comment by epoberezkin** [comment]
>  I'll have a look over these. Any amendments?

Not really, I've just made the spec a bit more specific in cases where wiki pages were a bit vague, but other than that the I implemented the proposals as they were written at the time.

> I mean with the same sentence structure etc, so its easy to later incorporate into the spec later and pickup for people who are familiar with the v4 spec. And so its fairly rigorous.

Got it. Yes, it makes sense to do it.

**Comment by sam-at-github** [comment]
Having a look over the wiki pages. It would be great if we could agree to follow the convention that where you create a wiki page for a v5 feature request / amendment you also link the title to a issue with the same title. This would make it easier to have and find the discussion on each proposal.

**Comment by brutzman** [comment]
Topic:  EXI for JSON draft released, inquiry about JSON Schema

At the World Wide Web Consortium (W3C), the Efficient XML Interchange (EXI) working group has published a first working draft of potential interest to this group.

Efficient XML Interchange (EXI) for JSON
W3C First Public Working Draft 28 January 2016
https://www.w3.org/TR/2016/WD-exi-for-json-20160128
Latest version: http://www.w3.org/TR/exi-for-json

Abstract. The Efficient XML Interchange (EXI) format is a compact representation that simultaneously optimizes performance and the utilization of computational resources. The EXI format was designed to support XML representation. With a relatively small set of transformations it may also be used for JSON, a popular format for exchange of structured data on the Web.

The EXI Working Group page is https://www.w3.org/XML/EXI

For those interested, discussion of the EXI for JSON working draft occurs on the public-exi@w3.org mailing list.
https://www.w3.org/XML/EXI/#Participation

We currently have interesting discussions about whether usage of JSON Schema together with EXI for JSON might lead to similar improvements in compaction and performance when using XML Schema with EXI. Further dialog will always be welcome.

Don Brutzman, Daniel Peintner and Taki Kamiya

**Comment by Acubed** [comment]
@brutzman I didn't know about this thanks, I've also had a task open for CBOR https://github.com/json-schema-org/json-schema-spec/issues/6

**Comment by tekwiz** [comment]
Given that v4 is expired but, for all intents and purposes, is a defacto standard (as @webron points out in https://github.com/OAI/OpenAPI-Specification/issues/590#issuecomment-190960733), does it make sense to initiate a standards effort to renew and formalize v5 as JSON Schema currently exists in the wild?  This would be similar to the initial work of the httpbis WG when they consolidated/clarified/renewed HTTP 1.1 (see the [httpbis Charter for Working Group](https://datatracker.ietf.org/wg/httpbis/charter/)).

**Comment by Relequestual** [comment]
I believe the current working plan is for v4 draft to become v1.0 first. Not sure on the status of this effort though.

**Comment by Acubed** [comment]
The point of this issue is that current v4 draft has some serious vocabulary issues that should exclude RFC publication; we want to fix these without changing much actual functionality from v4.

(Regular recurring reminder the current draft needs peer review, come take a look at https://github.com/json-schema-org/json-schema-spec please!)

**Comment by Lcfvs** [comment]
Hi,
How about multiple patterns support on a same field?

Some cases really need it, like robust password fields.

And it helps on debug, to cut a rules collection, to provide a specific error message.

Example:
```JS
    {
        type: String,
        required: true,
        minlength: 8,
        maxlength: 100,
        validate: [
            {
                validator: /[a-zA-Z]/,
                msg: 'password must contains a letter'
            },
            {
                validator: /\d/,
                msg: 'password must contains a number'
            },
            {
                validator: /[^a-zA-Z]/,
                msg: 'password must contains a special character'
            }
        ]
    }
```

**Comment by Acubed** [comment]
@Lcfvs This would be an new feature, and thus (under this proposal) an issue for v6, if v5 is to iron out known problems with the current draft and get a new expiry-date.

Note, however, you can accomplish roughly the same thing you're asking with the `"allOf"` property:

```javascript
{
    allOf: [
        {type:"string", description:"Must be a string"},
        {pattern:"\d", description:"Must contain a number"},
    ]
}
```

**Comment by the-t-in-rtf** [comment]
@Acubed, there is already a proposal to include custom error messages in v5:

https://github.com/json-schema/json-schema/wiki/Custom-error-messages-(v5-proposal)

What are the odds that will make it in?

**Comment by Lcfvs** [comment]
@Acubed, Ok, ok, waiting the next gen, then, because the allOf seems a few tricky.

@the-t-in-rtf, I like the idea about error messages, it improves the debug but imho, the best way to do it, is to directly associate constraint & related message, for readability and for multiples constraints of the same type. (like multiple patterns, for example)

**Comment by brettz9** [comment]
Could I get a reply on my i18n comment?

**Comment by the-t-in-rtf** [comment]
@Lcfvs, the proposed standard is pretty clear about associating messages with the specific rule being broken.  You can do it from inside the rule definition using the `errors` block in context, for example:

```
{
  "type": "string",
  "pattern": "^[A-Z]+$",
  "errors":  {
    "type": "You  must enter a string",
    "pattern": "The string must be UPPERCASE."
  }
}
```

You can also add rules from the document level using JSON pointers that refer to the deep location of the rule being broken (see the link for examples).  This seems ideal for creating derived schemas or creating schemas with error messages in multiple languages.

**Comment by Lcfvs** [comment]
@the-t-in-rtf, hum, yeah, but how about multiple patterns, for example?

How to identify what pattern mismatch?

For me, on rule, one related error.

**Comment by Acubed** [comment]
In the interest of keeping this issue on-topic, proposals and comments should be made at https://github.com/json-schema-org/json-schema-spec please (We really need peer review, by all means, comment there!)

**Comment by Lcfvs** [comment]
Ok, sorry. :|

**Comment by brettz9** [comment]
If you want to welcome comments, do you want to open the issue tracker there in addition to PRs?

**Comment by epoberezkin** [comment]
@acubed I think the important thing to do is to move the website to the new organisation (I see the fork but I am not sure if www.json-schema.org is hosted there already) and to change the link on the home page so it points to the new org. While it points here, there will be comments here...

I also think it would be good to add more owners and contributors than just you to the new org, to avoid the risk of repeating the same situation that caused this migration. At least the same people who are now in json-schema organisation.

Also, I've made a couple of PRs for the website previously, I guess I have to re-submit them there?

**Comment by Acubed** [comment]
@epoberezkin We might be able to handle those in the existing repo, someone with existing write permissions there would have to merge it in, but it's entirely possible

@geraintluff et el. Can we remove the `CNAME` file from the `gh-pages` branch?

**Comment by epoberezkin** [comment]
@Acubed thanks

**Comment by Relequestual** [comment]
Myself and @Acubed are only Members. Granted, ownership for myself is possibly unwranted with my current level of involvement, but I would be unable to abandon the project or leave it hanging.

I DID make some promises to do a big chunk of work for a new json-schema website. While the response was positive, further communication on how to proceed wasn't recpirpcated. I feel I was effectivly told to hold off till more work had been done to further to draft being made a spec.

I may come back to the effort over the next few months! We shall see.

**Comment by the-t-in-rtf** [comment]
Although I'm obviously looking forward to seeing the standard evolve, thanks very much to everyone who has stepped up to keep the standard alive.

**Comment by nemesisdesign** [comment]
I'm interested in contributing if you need some sort of help.

I'm currently involved in [NetJSON](http://netjson.org) a data interchange format that describes computer networks (network graph, configuration, monitoring) which is gaining some traction.
The (draft 0) [NetJSON RFC](http://netjson.org/rfc.html) currently mentions the JSON-Schema v4 draft and one implementation in particular makes [heavy use of JSON-Schema](https://github.com/openwisp/netjsonconfig/blob/master/netjsonconfig/schema.py).
NetJSON needs JSON-Schema for its implementation, so I want to give my contribution if possible.

I would really like to see one of these two proposals included in the v5:

* [choices (v5 proposal to enhance enum)](https://github.com/json-schema/json-schema/wiki/choices-(v5-proposal-to-enhance-enum))
* [enumNames (v5 proposal)](https://github.com/json-schema/json-schema/wiki/enumNames-(v5-proposal))

Although I would prefer `choices` because it can be considered a way to decorate the schema just like `title` and `description`.

**Comment by BigBlueHat** [comment]
:+1: to plans to add `contains`.

We're using JSON Schema (via `ajv`) in our testing platform for [Web Annotation Data Model](http://www.w3.org/TR/annotation-model/) and having `contains` would greatly clarify situations like [validating that a specific `@context` value is used](https://github.com/Spec-Ops/web-platform-tests/blob/master/annotation-model/common/contextValue.json#L15-L26).

**Comment by epoberezkin** [comment]
@BigBlueHat ajv supports contains btw

**Comment by BigBlueHat** [comment]
@epoberezkin yep! I found that here (and "reacted" to your note about it). Mostly just want to encourage it into the spec--and this seemed to be the right/only place to do that. :smile: Thanks for `ajv`!

**Comment by awwright** [comment]
@BigBlueHat @epoberezkin JSON-LD is defined in such a way that makes it difficult if not impossible to describe with JSON Schema, in a stateless fashion.

I'd be very interested in helping the TR, though.

Direct feature requests to the issue tracker at https://github.com/json-schema-org/json-schema-spec

**Comment by Relequestual** [comment]
@awwright  ahem https://github.com/json-ld/json-ld.org/blob/master/schemas/jsonld-schema.json

**Comment by awwright** [comment]
It was a long and hard journey frought with polar bears, airplanes, sand castles, broken N64 controllers, and a busted speaker, but...

We did it!

https://tools.ietf.org/html/draft-wright-json-schema-00
https://tools.ietf.org/html/draft-wright-json-schema-validation-00
https://tools.ietf.org/html/draft-wright-json-schema-hyperschema-00

CC @Relequestual @epoberezkin @handrews et al.

I'll just wait on one more thing, getting @geraintluff et al. to sign off on the new series formally replacing the old series.

**Comment by darshakthakore** [comment]
Is the new draft going to be discussed in the working group meetings? 
I didn't see it being listed in the jsonbis wg.

**Comment by awwright** [comment]
It has yet to be adopted by a WG. Idk if it's in-scope for _jsonbis_, but maybe someone else like _art_

**Comment by awwright** [comment]
The documents now formally supercede the old series, looks like Kris Zyp or the IETF Secretariat got around to approving it!
